transports/quic/src/transport.rs

```diff
@@ -0,0 +1,703 @@
+    dialer: HashMap<SocketFamily, Dialer>,
+
+    _marker: PhantomData<P>,
+}
+
+impl<P> GenTransport<P> {
+    pub fn new(config: Config) -> Self {
+        Self {
+            listeners: SelectAll::new(),
+            config,
+            dialer: HashMap::new(),
+            _marker: Default::default(),
+        }
+    }
+}
+
+/// Error that can happen on the transport.
+#[derive(Debug, thiserror::Error)]
+pub enum TransportError {
+    /// Error while trying to reach a remote.
+    #[error(transparent)]
+    Reach(#[from] quinn_proto::ConnectError),
+    /// Error after the remote has been reached.
+    #[error(transparent)]
+    Established(#[from] ConnectionError),
+
+    #[error(transparent)]
+    Io(#[from] std::io::Error),
+
+    /// The task driving the endpoint has crashed.
+    #[error("Endpoint driver crashed")]
+    EndpointDriverCrashed,
+}
+
+impl<P: Provider> Transport for GenTransport<P> {
+    type Output = (PeerId, QuicMuxer);
+    type Error = TransportError;
+    type ListenerUpgrade = Connecting;
+    type Dial = BoxFuture<'static, Result<Self::Output, Self::Error>>;
+
+    fn listen_on(
+        &mut self,
+        addr: Multiaddr,
+    ) -> Result<ListenerId, CoreTransportError<Self::Error>> {
+        let socket_addr = multiaddr_to_socketaddr(&addr)
+            .ok_or(CoreTransportError::MultiaddrNotSupported(addr))?;
+        let listener_id = ListenerId::new();
+        let listener = Listener::new::<P>(listener_id, socket_addr, self.config.clone())?;
+        self.listeners.push(listener);
+
+        // Remove dialer endpoint so that the endpoint is dropped once the last
+        // connection that uses it is closed.
+        // New outbound connections will use a bidirectional (listener) endpoint.
+        self.dialer.remove(&socket_addr.ip().into());
+
+        Ok(listener_id)
+    }
+
+    fn remove_listener(&mut self, id: ListenerId) -> bool {
+        if let Some(listener) = self.listeners.iter_mut().find(|l| l.listener_id == id) {
+            listener.close(Ok(()));
+            true
+        } else {
+            false
+        }
+    }
+
+    fn address_translation(&self, listen: &Multiaddr, observed: &Multiaddr) -> Option<Multiaddr> {
+        if !is_quic_addr(listen) || !is_quic_addr(observed) {
+            return None;
+        }
+        Some(observed.clone())
+    }
+
+    fn dial(&mut self, addr: Multiaddr) -> Result<Self::Dial, CoreTransportError<Self::Error>> {
+        let socket_addr = multiaddr_to_socketaddr(&addr)
+            .ok_or_else(|| CoreTransportError::MultiaddrNotSupported(addr.clone()))?;
+        if socket_addr.port() == 0 || socket_addr.ip().is_unspecified() {
+            return Err(CoreTransportError::MultiaddrNotSupported(addr));
+        }
+        let mut listeners = self
+            .listeners
+            .iter_mut()
+            .filter(|l| {
+                let listen_addr = l.endpoint_channel.socket_addr();
```

Comment:
Do you need to check if this listener is closed?

Reply:
Good catch!